src/backend/gsroa.c

```diff
@@ -49,15 +54,66 @@ static void sliceStructs_Gather(SymInfo *sia, elem *e)
                 if (si >= 0 && sia[si].canSlice)
                 {
                     assert(si < globsym.top);
-                    if (tysize(e->Ety) != REGSIZE ||
-                        (e->Eoffset != 0 && e->Eoffset != REGSIZE))
+                    unsigned sz = tysize(e->Ety);
+                    if (sz == 2 * REGSIZE)
+                    {
+                        // Rewrite as OPpair later
+                        sia[si].usePair = true;
+
+                        /* OPpair cannot handle XMM registers, cdpair() and fixresult()
+                         */
+                        if (tyfloating(sia[si].ty0) || tyfloating(sia[si].ty1))
```

Comment:
This branch is a bit confusing b/c it seems to rely on the fact that the `else if` branch below was run for `e->E1` and `e->E2` beforehand.
Would be simpler to understand if you changed the order of the branches and added a small comment.

Reply:
Thanks, just spent 2 days debugging a memory corruption based on this :/.
Wasn't too well put here, but as mentioned `ty0` and `ty1` are still uninitialized if access as fat value happens before any access to a sliced field.